std/node/assertion_error.ts

```diff
@@ -0,0 +1,585 @@
+        // them in case more than five lines were identical.
+        if (identical > 2) {
+          if (identical > 3) {
+            if (identical > 4) {
+              if (identical === 5) {
+                res += `\n  ${actualLines[i - 3]}`;
+                printedLines++;
+              } else {
+                res += `\n${blue}...${white}`;
+                skipped = true;
+              }
+            }
+            res += `\n  ${actualLines[i - 2]}`;
+            printedLines++;
+          }
+          res += `\n  ${actualLines[i - 1]}`;
+          printedLines++;
+        }
+        // No identical lines before.
+        identical = 0;
+        // Add the actual line to the result and cache the expected diverging
+        // line so consecutive diverging lines show up as +++--- and not +-+-+-.
+        res += `\n${green}+${white} ${actualLine}`;
+        other += `\n${red}-${white} ${expectedLine}`;
+        printedLines += 2;
+        // Lines are identical
+      } else {
+        // Add all cached information to the result before adding other things
+        // and reset the cache.
+        res += other;
+        other = "";
+        identical++;
+        // The very first identical line since the last diverging line is be
+        // added to the result.
+        if (identical <= 2) {
+          res += `\n  ${actualLine}`;
+          printedLines++;
+        }
+      }
+    }
+    // Inspected object to big (Show ~50 rows max)
+    if (printedLines > 50 && i < maxLines - 2) {
+      return `${msg}${skippedMsg}\n${res}\n${blue}...${white}${other}\n` +
+        `${blue}...${white}`;
+    }
+  }
+
+  return `${msg}${skipped ? skippedMsg : ""}\n${res}${other}${end}${indicator}`;
+}
+
+export interface AssertionErrorConstructorOptions {
+  message?: string;
+  actual?: unknown;
+  expected?: unknown;
+  operator?: string;
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  stackStartFn?: Function;
+  // Compatibility with older versions.
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  stackStartFunction?: Function;
+}
+
+export class AssertionError extends Error {
+  [key: string]: unknown
+
+  // deno-lint-ignore constructor-super
+  constructor(options: AssertionErrorConstructorOptions) {
+    if (typeof options !== "object" || options === null) {
+      throw new ERR_INVALID_ARG_TYPE("options", "Object", options);
+    }
+    const {
+      message,
+      operator,
+      stackStartFn,
+      // Compatibility with older versions.
+      stackStartFunction,
+    } = options;
+    let {
+      actual,
+      expected,
+    } = options;
+
+    // TODO: `stackTraceLimit` should be added to `ErrorConstructor` in
+    // cli/dts/lib.deno.shared_globals.d.ts
+    // deno-lint-ignore ban-ts-comment
+    // @ts-ignore
+    const limit = Error.stackTraceLimit;
+    // deno-lint-ignore ban-ts-comment
+    // @ts-ignore
+    Error.stackTraceLimit = 0;
```

Comment:
Ignoring all the lint mumbo jumbo - what is happening here? Why `Error.stackTraceLimit = 0;` ?

Reply:
The lint ignoration was added because TS doesn't know about the existence of `stackTraceLimit` property on type `ErrorConstructor`. It should be considered to declare it in `cli/dts/lib.deno.shared_globals.d.ts`. However I removed that since your review in favor of an [interface](https://github.com/schwarzkopfb/deno/blob/assertion_error/std/node/assertion_error.ts#L393) that extends `ErrorConstructor`.

`stackTraceLimit` is set to `0` to disable stack trace collection until `super()` gets called. After the invocation of parent constructor (`Error`) that limit gets [restored](https://github.com/schwarzkopfb/deno/blob/assertion_error/std/node/assertion_error.ts#L521).
The reason to do this is to avoid the overhead introduced by collecting stack frames which are [re-captured anyway](https://github.com/schwarzkopfb/deno/blob/assertion_error/std/node/assertion_error.ts#L536). We need to generate our own trace to be able to respect `stackStartFunction` and to manipulate error's name (include Node's error identifier in it).

This is necessary to mimic Node's behavior as close as possible and coming "as is" from the [original JS implementation](https://github.com/nodejs/node/blob/master/lib/internal/assert/assertion_error.js#L324).